File: bench/driver.js
// The TypeScript leg of the throughput suite. See README.md.
//
// Reads the built package, so `npm run build` in typescript/ has to have run.
// Prints one `task|nanos|per-second|batch|guard` line per task, in the order
// every other leg prints them.

const path = require('path');

const main = process.env.GPC_TYPESCRIPT_MAIN
    || path.join(__dirname, '..', 'typescript', 'dist', 'index.js');
const { GPC } = require(main);

// ---------------------------------------------------------------------------
// The same inputs as every other leg, built by the same integer arithmetic so
// that no data file has to be shared to make the four measure the same work.

const COUNT = 256;
const MASK = COUNT - 1;

const LATITUDES = [];
const LONGITUDES = [];
for (let i = 0; i < COUNT; i += 1) {
    LATITUDES.push(((i * 7919) % 17000) / 100 - 85);
    LONGITUDES.push(((i * 6271) % 35000) / 100 - 175);
}

const CODES = LATITUDES.map((latitude, i) => GPC.encode(latitude, LONGITUDES[i]));
const BARE = CODES.map((code) => code.replace(/[^0-9A-Z]/g, ''));
const GRID = BARE.map((bare) => GPC.codeToGrid(bare));
const ROWS = GRID.map((pair) => pair[0]);
const COLUMNS = GRID.map((pair) => pair[1]);
const MESSY = CODES.map((code) => code.toLowerCase());

// ---------------------------------------------------------------------------
// Whole batches rather than single operations: per-operation call overhead is
// a large part of what would otherwise be timed, and it is not what is being
// asked about. Each returns a number summed from every operation, because an
// engine may delete work whose result nothing looks at.

function taskEncode(size) {
    let total = 0;
    for (let i = 0; i < size; i += 1) {
        const j = i & MASK;
        total += GPC.encode(LATITUDES[j], LONGITUDES[j]).length;
    }
    return total;
}

function taskDecode(size) {
    let total = 0;
    for (let i = 0; i < size; i += 1) {
        const [latitude, longitude] = GPC.decode(CODES[i & MASK]);
        total += latitude + longitude;
    }
    return total;
}

function taskGridToCode(size) {
    let total = 0;
    for (let i = 0; i < size; i += 1) {
        const j = i & MASK;
        total += GPC.gridToCode(ROWS[j], COLUMNS[j]).length;
    }
    return total;
}

function taskCodeToGrid(size) {
    let total = 0;
    for (let i = 0; i < size; i += 1) {
        const [row, column] = GPC.codeToGrid(BARE[i & MASK]);
        total += row + column;
    }
    return total;
}

function taskNormalise(size) {
    let total = 0;
    for (let i = 0; i < size; i += 1) {
        total += GPC.normalise(MESSY[i & MASK])[0].length;
    }
    return total;
}

/**
 * Arithmetic the library has nothing to do with.
 *
 * Every figure is divided by this one, so a runner having a slow morning is
 * slow at both and the ratio survives the shared machine that the absolute
 * number does not.
 */
function taskCalibrate(size) {
    let total = 0;
    for (let i = 0; i < size; i += 1) {
        let x = i;
        for (let k = 0; k < 64; k += 1) {
            x = (Math.imul(x, 1103515245) + 12345) & 0x7FFFFFFF;
        }
        total += x;
    }
    return total;
}

const TASKS = [
    ['encode', taskEncode],
    ['decode', taskDecode],
    ['gridToCode', taskGridToCode],
    ['codeToGrid', taskCodeToGrid],
    ['normalise', taskNormalise],
    ['calibrate', taskCalibrate],
];

// ---------------------------------------------------------------------------

const LONG_ENOUGH = 50000000n;      // nanoseconds a batch must last to be timed
const WARMUP = 5;
const ROUNDS = 7;
const CAP = 1 << 24;

let guard = 0;

function measure(task, size) {
    const started = process.hrtime.bigint();
    const total = task(size);
    const elapsed = process.hrtime.bigint() - started;
    guard += total;
    return elapsed;
}

function run(name, task) {
    // Grow the batch until the clock's resolution stops being part of the
    // answer. The growing doubles as warm-up, which matters more here than in
    // most places: this code starts interpreted and is compiled only once it
    // has been run enough to look worth compiling.
    let size = 64;
    while (measure(task, size) < LONG_ENOUGH && size < CAP) size *= 2;

    for (let i = 0; i < WARMUP; i += 1) measure(task, size);

    const batches = [];
    for (let i = 0; i < ROUNDS; i += 1) batches.push(measure(task, size));
    batches.sort((one, other) => (one < other ? -1 : one > other ? 1 : 0));

    const nanos = Number(batches[(ROUNDS - 1) >> 1]) / size;
    process.stdout.write(
        `${name}|${nanos.toFixed(4)}|${(1e9 / nanos).toFixed(0)}|${size}|${guard.toFixed(0)}\n`,
    );
}

for (const [name, task] of TASKS) run(name, task);

File: bench/Driver.java
// The Java leg of the throughput suite. See README.md.
//
// Prints one `task|nanos|per-second|batch|guard` line per task, in the order
// every other leg prints them. Compiled against the classes next door by
// bench/bench.py, the same way the conformance driver is.

import ca.pranavpatel.algo.gridpointcode.Coordinates;
import ca.pranavpatel.algo.gridpointcode.GPC;

import java.util.Arrays;
import java.util.function.IntToDoubleFunction;

public class Driver {

    // -----------------------------------------------------------------------
    // The same inputs as every other leg, built by the same integer arithmetic
    // so the four measure the same work with no data file between them.

    static final int COUNT = 256;
    static final int MASK = COUNT - 1;

    static final double[] LATITUDES = new double[COUNT];
    static final double[] LONGITUDES = new double[COUNT];
    static final String[] CODES = new String[COUNT];
    static final String[] BARE = new String[COUNT];
    static final long[] ROWS = new long[COUNT];
    static final long[] COLUMNS = new long[COUNT];
    static final String[] MESSY = new String[COUNT];

    static {
        for (int i = 0; i < COUNT; i++) {
            LATITUDES[i] = ((i * 7919L) % 17000) / 100.0 - 85.0;
            LONGITUDES[i] = ((i * 6271L) % 35000) / 100.0 - 175.0;
            CODES[i] = GPC.Encode(LATITUDES[i], LONGITUDES[i]);
            BARE[i] = CODES[i].replaceAll("[^0-9A-Z]", "");
            long[] grid = GPC.CodeToGrid(BARE[i]);
            ROWS[i] = grid[0];
            COLUMNS[i] = grid[1];
            MESSY[i] = CODES[i].toLowerCase();
        }
    }

    // -----------------------------------------------------------------------
    // Whole batches rather than single operations: at these speeds a call per
    // operation is a large part of what would be timed and is not what is being
    // asked about. Each returns a number summed from every operation, because a
    // just-in-time compiler may delete work whose result nothing reads.

    static double encode(int size) {
        double total = 0;
        for (int i = 0; i < size; i++) {
            int j = i & MASK;
            total += GPC.Encode(LATITUDES[j], LONGITUDES[j]).length();
        }
        return total;
    }

    static double decode(int size) {
        double total = 0;
        for (int i = 0; i < size; i++) {
            Coordinates point = GPC.Decode(CODES[i & MASK]);
            total += point.Latitude + point.Longitude;
        }
        return total;
    }

    static double gridToCode(int size) {
        double total = 0;
        for (int i = 0; i < size; i++) {
            int j = i & MASK;
            total += GPC.GridToCode(ROWS[j], COLUMNS[j]).length();
        }
        return total;
    }

    static double codeToGrid(int size) {
        double total = 0;
        for (int i = 0; i < size; i++) {
            long[] grid = GPC.CodeToGrid(BARE[i & MASK]);
            total += grid[0] + grid[1];
        }
        return total;
    }

    static double normalise(int size) {
        double total = 0;
        for (int i = 0; i < size; i++) {
            total += GPC.Normalise(MESSY[i & MASK])[0].length();
        }
        return total;
    }

    /**
     * Arithmetic the library has nothing to do with.
     *
     * Every figure is divided by this one, so a runner having a slow morning is
     * slow at both and the ratio survives the shared machine that the absolute
     * number does not.
     */
    static double calibrate(int size) {
        double total = 0;
        for (int i = 0; i < size; i++) {
            int x = i;
            for (int k = 0; k < 64; k++) {
                x = (x * 1103515245 + 12345) & 0x7FFFFFFF;
            }
            total += x;
        }
        return total;
    }

    // -----------------------------------------------------------------------

    static final long LONG_ENOUGH = 50_000_000L;
    static final int WARMUP = 5;
    static final int ROUNDS = 7;
    static final int CAP = 1 << 24;

    static double guard = 0;

    static long measure(IntToDoubleFunction task, int size) {
        long started = System.nanoTime();
        double total = task.applyAsDouble(size);
        long elapsed = System.nanoTime() - started;
        guard += total;
        return elapsed;
    }

    static void run(String name, IntToDoubleFunction task) {
        // Grow the batch until the clock's resolution stops being part of the
        // answer. The growing doubles as warm-up, which matters here more than
        // anywhere: this code is interpreted until it has been run enough times
        // to be worth compiling, and the difference is an order of magnitude.
        int size = 64;
        while (measure(task, size) < LONG_ENOUGH && size < CAP) size *= 2;

        for (int i = 0; i < WARMUP; i++) measure(task, size);

        long[] batches = new long[ROUNDS];
        for (int i = 0; i < ROUNDS; i++) batches[i] = measure(task, size);
        Arrays.sort(batches);

        double nanos = (double) batches[(ROUNDS - 1) / 2] / size;
        System.out.printf("%s|%.4f|%.0f|%d|%.0f%n",
                name, nanos, 1e9 / nanos, size, guard);
    }

    public static void main(String[] args) {
        run("encode", Driver::encode);
        run("decode", Driver::decode);
        run("gridToCode", Driver::gridToCode);
        run("codeToGrid", Driver::codeToGrid);
        run("normalise", Driver::normalise);
        run("calibrate", Driver::calibrate);
    }
}
